### ISSUE TITLE
TDbasedUFEadv

### ISSUE THREAD

**Issue by tagtag** [Bioconductor/Contributions#2944]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/tagtag/TDbasedUFEadv

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
On initial testing I am seeing 
```
--- re-building ‘Enrichment.Rmd’ using rmarkdown
Quitting from lines 41-68 (Enrichment.Rmd) 
Error: processing vignette 'Enrichment.Rmd' failed with diagnostics:
R: geometry does not contain image `/tmp/RtmplX2cjc/Rbuild5cdc763a79ea2/TDbasedUFEadv/vignettes/Enrichment_files/figure-html/unnamed-chunk-2-1.png' @ warning/attribute.c/GetImageBoundingBox/247

```

**Comment by vjcitn** [comment]
I don't know why I encountered that with R CMD check, but I will assume it will be remedied if it occurs on the build system.  I noted many spelling errors in vignettes; perhaps you can use an automated spelling analyzer to deal with this.  The QuickStart has `we continue to type 1 and press enter unril we can see the nineth one as` -- this approach to exploring data is outmoded.  The package passes check so I will pass it to review, but I would suggest you have a look at a) curatedTCGAData as a way of compactly managing many TCGA tumors, and b) shiny as a way of managing user interactions with a large number of results.  Developer mentoring may be available to discuss how a shiny app might help your users.

**Comment by tagtag** [comment]
@vjcitn Thank you very much for your comments. The error did not occur in our  system. In the case we cannot correct it, I will ask you. I will check the spelling. I will consider the usage of curatedTCGAData as well as shiny.

**Comment by tagtag** [comment]
@lazappi Although @vjcitn advised me to use curatedTCGAData instead of RTCGA, I found that it is not suitable for us to employ curatedTCGAData. In our vignettes QuickStart, we used all TCGA cancer data set from RTCGA.rnaseq. If we try to do the same by curatedTCGAData, we need to perform
 all <- curatedTCGAData(
    "*", "RNASeq2Gene", version = "2.0.1", dry.run = FALSE
)
and curatedTCGAData started to download data. Download did not finish even after one hour passed. In contrast to curatedTCGAData, RTCGA.rnaseq includes data within it, thus no need to download. Then I gave up the usage of curatedTCGAData. If you think that I am wrong, I am glad if you can advise me the correct way of using curatedTCUAData.

**Comment by tagtag** [comment]
@lazappi @vjcitn also advised me to use shiny instead of menu. However, I found that bioconductor can only accept shiny as a whole package, not a part of package.
https://contributions.bioconductor.org/shiny.html
TDbasedUFEadv cannot be shiny app as a whole, since only a part of package is interactive. Thus, the usage of shiny as a part of TDbasedUFE is impossible. If you think that I am wrong, I am glad if you can advise me the correct way of using shiny in TDbasedUFEadv.

**Comment by lazappi** [comment]
- I'm not familiar with the TCGA packages, can you explain the difference between **{curatedTCGAData}** and **{RTCGA}**?
- What do you mean by "can only accept shiny as a whole package"? I couldn't see a reference to that in the documentation you linked to and if fact I think it requires that a shiny app functions as a normal package with only the interactive part being included in the app.
- Please address the errors from the build system before I can review the package.

**Comment by vjcitn** [comment]
Let's leave curatedTCGAData to one side for now; @LiNk-NY may wish to comment at some point but it is not urgent.  This comment in the linked element of contributors.bioconductor.org `[Shiny apps](https://shiny.rstudio.com/) can be submitted to Bioconductor as software packages.` seems misleading to me.  Packages may include shiny apps but the apps are supposed to be testable functions.  I will address this with the maintainers of the guidelines.  I think it is overbroad wording.

**Comment by tagtag** [comment]
@vjcitn @lazappi Thanks for your valuable comments. Yes, I postpone the comments about curatedTCGA at the moment.  As for Shiny, I am waiting for the feedback from you or maintainers to know if menu functions I implemented can be replaced with Shiny apps included in individual packages, not as an independent package that includes only Shiny apps. As for errors, I will check it but I was also asked to check spelling. After spell check I will also address build error (actually speaking, we do not have this errors in our side but I will try to fix it).

**Comment by tagtag** [comment]
@vjcitn @lazappi Thanks for your valuable comments. Yes, I postpone the comments about curatedTCGA at the moment.  As for Shiny, I am waiting for the feedback from you or maintainers to know if menu faunctions can be replaced with Shiny apps included in individual packages, not as an independent package that includes only Shiny apps. As for errors, I will check it but I was also asked to check spelling. After spell check I will also adress build error (actually speaking, we do not have this errors in outr side but I will try to fix it).

**Comment by LiNk-NY** [comment]
> all <- curatedTCGAData(
> "*", "RNASeq2Gene", version = "2.0.1", dry.run = FALSE
> )
> and curatedTCGAData started to download data.

Hi Taguchi, @tagtag 

I noticed that the `assays` argument (`RNASeq2Gene`) would return both `RNASeq2Gene` and `RNASeq2GeneNorm`. Was this the intention? Perhaps, but in either case, a sole `RNASeq2Gene` input should only return `RNASeq2Gene` type of assays. I have updated `curatedTCGAData` to work this way (see version `1.21.2`). If you _do_ intend to work with both type of assays, the input should be `RNASeq2Gene*`. 

Note that `curatedTCGAData` software is lightweight and only downloads the data requested. If you are requesting two assay types from all cancer types, that would require the download of 69 assay files and 66 metadata files (which include `colData` and `metadata`) **versus** 33 assay files and 66 metadata files when requesting only `RNASeq2Gene` assays. 

The download time would also depend on your connection speed relative to the hosting location.

**Comment by tagtag** [comment]
@LiNk-NY I know that my code download two assays, (thus in total 69 files), and they can be 33 files. However, even if 33 files are download, it took so long time (It is not lightweighted if we need to download as many as 33 files).  I know that it depends upon connection speed, too, but I dislike to use something related to connection speed in my package, but something store data in local PC like RTCGA. I am glad if you can consider the situation here. Is it acceptable? > @vjcitn @lazappi

**Comment by LiNk-NY** [comment]
The typical use case for `curatedTCGAData` involves a single cancer type and is quite lightweight and integrative. 

> Is it acceptable?

If I may, I think the use of `RTCGA.rnaseq` is acceptable. I would suggest making it clear in the vignette that you are using old release TCGA data from 20151101 (based on the package version number and as shown on the website https://rtcga.github.io/RTCGA/).

**Comment by tagtag** [comment]
@LiNk-NY Thanks. I will mention about it in Vignettes. Since I have used RTCGA only for demonstration purpose in Vignettes, it does not matter that it is an old package.

**Comment by tagtag** [comment]
The build error caused seems to be due to system errors, not because of my package, since it says

>Could not find executable pandoc-citeproc

This means that pandoc is missing in the system that tries to build.
I will postpone the effort to address other issues until this problem is fixed.

**Comment by tagtag** [comment]
@LiNk-NY Although I have asked about this warning on bioc-devel-ml, since I was advised to continue the discussion here, in this issue, I ask you the solution.

My query:

_I have gotten the following warning.
"WARNING: R CMD check exceeded 10 min requirement"
[[History]Bioconductor Single Package Builder](http://bioconductor.org/spb_reports/TDbasedUFEadv_buildreport_20230307103513.html#merida1_check_anchor)
Does it mean that I need to fragment the package into smaller ones?_


Answer from bioc-devel-ml:

_Please continue this conversation on the issue being reviewed._  

_In general, packages are expected to build and check in a reasonable amount of time.  If you see this across all platform, there may be opportunities to improve your code and make it more efficient.  If it is a single platforms we can evaluate on a case by case basis as one of the builders may be experiencing a heavy load that could affect this result._


My case seems to correspond to the case "it is a single platforms"  since Check in UNIX seems to be OK

![image](https://user-images.githubusercontent.com/11694497/223477463-b462039c-fc47-4f50-905f-fbaedcb21406.png)

So, what is the next action that I can take? I am glad if you can advice me about it.

**Comment by lazappi** [comment]
I think that is a reasonable solution, but maybe you could also mention that **{curatedTCGAData}** can be used to get the more recent data if the user wants to.

For the Shiny question, I don't think there is any advice that says a package can't include a Shiny app along with other things so I think that should be fine. I haven't looked at exactly what you want to do with it though so we could discuss your concerns a bit more if you want.

**Comment by tagtag** [comment]
@lazappi Thanks for your valuable comments. 

> you could also mention that {curatedTCGAData} can be used to get the more recent data if the user wants to.

Yes, I did in updated vignettes.

> we could discuss your concerns a bit more if you want.

Actually, it is not my concern.  Although I have used menu() function, @vjcitn advised me that it is  an old fashioned way and it is better to use Shiny. I used menu() function to let users to select one plot among multiple plots (the number of plots to be shown vary, thus fixed layout, e.g. 3 x 2, is not a good idea). I showed plots one by one, user is forced to select the best one. I offered the three choices,  next, previous, select. Then users used next and previous to investigate plot and finally select one of them by select. Do you think that it is better to implement this function with Shiny?   If the fixed layout, e.g., 2 x3 can be employed, Shiny is a good idea since it can allow users to select one of them at a glace  (although 2x3 layout can be also employed with menu() if we can use mfrow=c(2,3) and accepd number as input for menu(). )

**Comment by lazappi** [comment]
The build time is most likely due to either the vignette or function example. I would see if you can make these a bit quicker, the simplest thing to try is probably reducing the size of any example datasets used.

I'm not very familiar with using menus but I am happy to take the advice from @vjcitn. Please look into whether you can use Shiny instead.

**Comment by tagtag** [comment]
@LiNk-NY As for build time,  UNIX has no problem and only mac has. In that case, I was informed that "we can evaluate on a case by case basis as one of the builders may be experiencing a heavy load that could affect this result." I am glad if you can tell me who can do this. 

As for Shiny, I am not sure if Shiny is fitted to my purpose, since  the number of plots that I have to show the user  and user must select one among is unlimited. In that case, I think that Shiny is not fitted. Shiny interface is very fitted only to select something among choices that can be shown in a single page. What is your ompinion?

**Comment by LiNk-NY** [comment]
@tagtag 
It would be good to know, what is taking most of the nearly 15 minutes on the Mac builder. 
If it were closer to 10 minutes, it would be okay. 
Note that the examples take less than 2 minutes: 
```r
Examples with CPU (user + system) or elapsed time > 5s
                         user system elapsed
selectFeatureTransRect 55.422  0.569  65.380
prepareCondTCGA        12.198  0.895  15.169
```
Perhaps it is code in the vignette, unit tests, or code in the package? 

In general, it would be advisable to reduce the time the package takes to check by using smaller exemplar datasets as Luke @lazappi mentioned.

**Comment by tagtag** [comment]
@LiNk-NY Possibly, it is in vignettes. Since it is difficult to reduce time spent by vignette because we need to show the user the validity of the method that requires substantial cpu time, is it acceptable not to execute R code in vignettes on build, but just include the output from R code?  I think that it is possible if we can replace executable chunks with not executable ones.

**Comment by LiNk-NY** [comment]
The vignette should demonstrate the functionality of the package with minimal examples. It is not meant to showcase a cpu intensive validation. For the latter purpose, please create a workflow package and/or move the code to a separate repository (e.g., manuscript preparation repository). 

> is it acceptable not to execute R code in vignettes on build, but just include the output from R code?

It is not acceptable because code in the vignette should be functional and tested; otherwise, it can go stale.

**Comment by tagtag** [comment]
@LiNk-NY What is "manuscript preparation repository"? I have never heard about it in Bioconductor.

**Comment by LiNk-NY** [comment]
@tagtag 
It seems like the files that you have in the vignette are part of an academic publication. Is that the case? If so, the figures and the code that generates the figures should go in a separate personal repository on GitHub. This is what I mean by "manuscript preparation repository."

**Comment by tagtag** [comment]
@LiNk-NY No, it doesn't. It is original research I have developed for this vignettes.  Thus, if we reduce the size of data, none can be confident of my method. Since UNIX can process R CMD check within limited time, is it possible for you to tell me how to pass it in Mac, too? PS I do not have any Mac box, thus I cannot check it by my self.

**Comment by tagtag** [comment]
@LiNk-NY I have found that too long time for R CMD check in Mac is universal, not only for mine.
https://github.com/elsasserlab/labcode/issues/63
I believe that it is unrealistic to ask developer to consider all platforms some of which have their own problem. How about your opinion?

**Comment by lazappi** [comment]
I am going to agree with what @LiNk-NY said

- The purpose of a vignette is to demonstrate how to use the package, not to necessarily show how effective a method is (that can be done in a separate paper/repository you can link to). There are plenty of packages that use small, sometimes unrealistic, datasets in their vignettes.
- One of the reasons Bioconductor is trusted is that packages are (mostly) tested on multiple platforms. It is likely you will have users who use MacOS so it's in your interest to try and make the performance similar to what you see on Unix.
- For both of these reasons it is important that code is run on the build system so we don't allow pre-computed results in vignettes (or at least not only that, it may be possible to run the code on an example dataset and then load the full results to use in later steps)
- As this is just a warning I think we can continue with the review (if you have completed the other changes we already discuss) but I think it is worth trying to address it

**Comment by tagtag** [comment]
@LiNk-NY @lazappi I will take the action that "it may be possible to run the code on an example dataset and then load the full results to use in later steps".  Thanks.

**Comment by tagtag** [comment]
@lazappi Now cpu time for R CMD check for mac was shorten so as to be within ten minutes. The remaining warning did not come from my side but from RITAN.
Warning: replacing previous import 'ensembldb::keys' by 'hash::keys' when loading 'RITAN'
Warning: replacing previous import 'ensembldb::filter' by 'stats::filter' when loading 'RITAN'
http://bioconductor.org/checkResults/devel/bioc-LATEST/RITAN/nebbiolo1-checksrc.html
which I am not responsible for.  I am glad if you can advise me how to pass check under this circumstances. I have no way to correct this waring, since it came from RITAN.  Or do I need to replace RITAN with something not associated with Warnings?

**Comment by tagtag** [comment]
@lazappi I have removed RITAN from TDbasedUFEadv to avoid warnings. If you can tell me how to include RITAN that has warnings, I am happy to recover RITAN. If not, please review the present version, 0.99.6, and let me know your opinion.

**Comment by lazappi** [comment]
Hi @tagtag

Thanks for submitting **TDbasedUFEadv** :tada:! Below is my review of your package. Please reply here if anything is unclear or needs any further explanation.

**What next?**

Please address the comments as best as you can. When you are ready for me to check the package again please reply to let me know with a summary of changes you have made or any other responses. You can use the "Quote reply" option in the `...` menu on this comment to reply directly to my points below.

Luke

# Review

**Key:** :rotating_light: Required :warning: Recommended :green_circle: Optional :question: Question

## General comments

- [ ] :question: The functions in the package seem to mostly be small helper functions that build on the functionality in **{TDbasedUFE}**. Have you considered adding them to the existing package rather than creating a new one?

## DESCRIPTION file

- [ ] :warning: It may be appropriate to add "Software" to the `biocViews` field
- [ ] :green_circle: It is recommended to add a `BugReports` field, this is usually a link to the GitHub issues page or the Bioconductor support site
- [ ] :green_circle: It is recommended to add a `URL` field, this is usually a link to the GitHub repository

## NAMESPACE file

- [ ] :rotating_light: `PrepareSummarizedExperimentTensorRect` should start with a lower case letter to match the other functions
- [ ] :rotating_light: Please use selective imports using `importFrom` instead of importing all with `import`

## The NEWS file

- [ ] :warning: Please keep the NEWS file up to date with the latest changes

## Package data

- [ ] :rotating_light: The data in `inst/extdata/` should be documented in `inst/script`

## Documentation

- [ ] :green_circle: There were quite a lot of typos in different places, the **{spelling}** package can help with finding these

### README

- [ ] :rotating_light: The `README` looks incomplete, please update to fill in the missing content
- [ ] :warning: It doesn't look as though you are using any RMarkdown functionality, if so it will be simpler to use a plain `README.md` file instead of `README.Rmd`

### Vignette

- [ ] :rotating_light: Please expand the introduction in the main vignette to describe the package in more detail. This should include the motivation for including the package in Bioconductor and comparisions to any similar packages (if applicable).
- [ ] :warning: It was not clear to me which order to look at the vignettes. I expected to start with `TDbasedUFEadv.Rmd` but that only convered technical detail, not any information about how to use the package. I suggest renaming/reorganising the vignettes to make this clearer.
- [ ] :warning: It also wasn't clear to me what the difference was between `Quickstart.Rmd` and `Quickstart2.Rmd`. I would suggest that this can be combined.
- [ ] :rotating_light: Whatever you decide to do I suggest using a more specific file name than "QuickStart.Rmd", sometimes it causes issues if different packages have vignettes with the same name
- [ ] :rotating_light: The main vignette should have an installation section that includes the Bioconductor installation instructions. These chunk should **not** be evaluated.
- [ ] :warning: There are lots of images in the vignettes directory, are all of these required or can some be produced by the vignette (I think this is related to the point about interactivity below)?
- [ ] :warning: I don't think you need to show all the menu items, I think it would be enough to mention there is an interactive option and maybe show one or two screenshots
- [ ] :warning: Please don't use `fig.keep="none"` if it can be avoided, in general the vignette should show the same output as the user would see if they ran the code themselves
- [ ] :green_circle: Usually it is better to use `library(package)` rather than `require(package)`. I also find it easier to follow the code if the library calls are in a separate chunk at the start.

### Man pages

- [ ] :rotating_light: You have included the word "Title" in the function documentation title which is then rendered in the final help page, either use `@title TEXT` or just `TEXT` on the first line
- [ ] :green_circle: Some of the variables in the examples are not very descriptive (e.g. `dummy`) which makes them a bit hard to understand

## Unit tests

- [ ] :rotating_light: Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks

## Code

### R

- [ ] :rotating_light: Please use named indicies instead of numeric indices where possible
- [ ] :warning: I found the `SummarizedExperimentTensorRect` class a bit unclear in a few ways
  - [ ] Given the name I expected it to inherit from `SummarizedExperiment` but that is not the case, I would suggest using a more descriptive name
  - [ ] It doesn't seem to be used by the user, how necessary is it to create a class rather than just using a list?
  - [ ] It is defined in a file called `init.R` which is not clear, usually this file is called something like `AllClasses.R` or the name of the class
- [ ] :warning: Please consider using more descriptive argument names, some of them (e.g. "k", "j", "LIST", "L") do not clearly describe what they are used for
- [ ] :warning: Consider adding checks for whether valid inputs are provided to function arguments
- [ ] :warning: I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`
- [ ] :warning: Generally functions should not output plots as side effects, generally it is better to have separate analysis and plotting functions
- [ ] :warning: It may be better to separate interactive functions for the non-interactive versions (or change how the options work), I found it a bit surprising that interactive was the default mode. Maybe one function could produce a plot that the user uses to select a parameter which they then pass to another function?
- [ ] :green_circle: This is mostly personal preference but I found your code quite hard to read because of the lack of indenting and whitespace. You may want to look at the [tidyverse style guide](https://style.tidyverse.org/) or the `styler` package.

## Shiny apps

- [ ] :question: I thought the plan was to replace the menus with a Shiny app, is this still the case?

**Comment by tagtag** [comment]
@LiNk-NY  Thank you very much for your detailed comments and efforts.

> ## General comments
> * [ ]  ❓ The functions in the package seem to mostly be small helper functions that build on the functionality in **{TDbasedUFE}**. Have you considered adding them to the existing package rather than creating a new one?

I can add them to existing package instead of generating new package.  In that case, they are not reviewed and  are distributed without reviews. Do you think that it is better to add them to existing package than generating new package?

**Comment by tagtag** [comment]
@LiNk-NY  I am glad, if you can let me know about your opinions in the following.

> * [ ]  ⚠️ There are lots of images in the vignettes directory, are all of these required or can some be produced by the vignette (I think this is related to the point about interactivity below)?

Yes, it is. Since interactive mode is not allowed to execute in installation, it is impossible for vignettes to generate figures  during installation.

> * [ ]  ⚠️ I don't think you need to show all the menu items, I think it would be enough to mention there is an interactive option and maybe show one or two screenshots

I do not think so. Batch mode will be never used by users, since batch is dummy to pass installation. Batch mode is empty.

> * [ ]  ⚠️ Please don't use `fig.keep="none"` if it can be avoided, in general the vignette should show the same output as the user would see if they ran the code themselves

As mentioned in the above, since interactive mode is not allowed, it is impossible for vignettes to generate figures during interactive mode.
 

> * [ ]  ⚠️ I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`

I have tried to avoid to use "for" as much as possible. Present "for" cannot be removed within my ability unless some one can shoe me how to replace them with "vapply" or "lapply" specifically.

> * [ ]  ⚠️ Generally functions should not output plots as side effects, generally it is better to have separate analysis and plotting functions

Since it is required to function as interactive mode, it is impossible to supress plot during function

> * [ ]  ⚠️ It may be better to separate interactive functions for the non-interactive versions (or change how the options work), I found it a bit surprising that interactive was the default mode. Maybe one function could produce a plot that the user uses to select a parameter which they then pass to another function?

As mentioned in the above, batch mode is empty. It does not do anything. Thus it is impossible for batch to be separate one.

> 
> ## Shiny apps
> * [ ]  ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?

Is it possible for you to check is Shiny is better than menu in this implementation? In my menu, figures are shown one by one with interaction with users. I do not think that Shiny is suitable for this purpose, although I agree with that menu is an old fashioned interface.

**Comment by lazappi** [comment]
> @LiNk-NY Thank you very much for your detailed comments and efforts.
> 
> > ## General comments
> > 
> > * [ ]   ❓ The functions in the package seem to mostly be small helper functions that build on the functionality in **{TDbasedUFE}**. Have you considered adding them to the existing package rather than creating a new one?
> 
> I can add them to existing package instead of generating new package. In that case, they are not reviewed and are distributed without reviews. Do you think that it is better to add them to existing package than generating new package?

In the end, this is up to you as the developer who has to maintain the packages but in general it is ok to add functionality to existing packages (unless someone from the core wants to correct me). There are some cases where I think it would be better to make a new package: if it exists a package maintained by somebody else, if there are significant dependencies that you don't want to add to the main package, if for some reason it makes maintenance easier. I'm not sure any of those apply in this case though.

> > * [ ]   ⚠️ There are lots of images in the vignettes directory, are all of these required or can some be produced by the vignette (I think this is related to the point about interactivity below)?
> 
> Yes, it is. Since interactive mode is not allowed to execute in installation, it is impossible for vignettes to generate figures during installation.
> 
> > * [ ]   ⚠️ I don't think you need to show all the menu items, I think it would be enough to mention there is an interactive option and maybe show one or two screenshots
> 
> I do not think so. Batch mode will be never used by users, since batch is dummy to pass installation. Batch mode is empty.
> 
> > * [ ]   ⚠️ Please don't use `fig.keep="none"` if it can be avoided, in general the vignette should show the same output as the user would see if they ran the code themselves
> 
> As mentioned in the above, since interactive mode is not allowed, it is impossible for vignettes to generate figures during interactive mode.
> 
> > * [ ]   ⚠️ I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`
> 
> I have tried to avoid to use "for" as much as possible. Present "for" cannot be removed within my ability unless some one can shoe me how to replace them with "vapply" or "lapply" specifically.
> 
> > * [ ]   ⚠️ Generally functions should not output plots as side effects, generally it is better to have separate analysis and plotting functions
> 
> Since it is required to function as interactive mode, it is impossible to supress plot during function
> 
> > * [ ]   ⚠️ It may be better to separate interactive functions for the non-interactive versions (or change how the options work), I found it a bit surprising that interactive was the default mode. Maybe one function could produce a plot that the user uses to select a parameter which they then pass to another function?
> 
> As mentioned in the above, batch mode is empty. It does not do anything. Thus it is impossible for batch to be separate one.
> 
> > ## Shiny apps
> > 
> > * [ ]   ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?
> 
> Is it possible for you to check is Shiny is better than menu in this implementation? In my menu, figures are shown one by one with interaction with users. I do not think that Shiny is suitable for this purpose, although I agree with that menu is an old fashioned interface.

I had not realised the batch mode is not functional and only there to pass the tests. This is something I think we want to avoid. It could be very confusing to the user and means the functions are not tested properly. I am beginning to think that Shiny is the best option in this case. That would give a nicer interface to the user than the menus and would make it easier to separate the processing code and the interface so that they can be checked properly.

**Comment by tagtag** [comment]
@LiNk-NY 

> I had not realised the batch mode is not functional and only there to pass the tests. This is something I think we want to avoid. It could be very confusing to the user and means the functions are not tested properly. I am beginning to think that Shiny is the best option in this case. That would give a nicer interface to the user than the menus and would make it easier to separate the processing code and the interface so that they can be checked properly.

What do you mean? Even if we are using Shiny, it is impossible to include  figures generated from interactive mode, we have to put figures generated from Shiny outside vignettes. Am I wrong? Or if I use Shiny, figures generated from Shiny can be included in vignettes automatically? I am very confused.  What is improved if we employ Shiny instead of menu?

**Comment by lazappi** [comment]
> > I had not realised the batch mode is not functional and only there to pass the tests. This is something I think we want to avoid. It could be very confusing to the user and means the functions are not tested properly. I am beginning to think that Shiny is the best option in this case. That would give a nicer interface to the user than the menus and would make it easier to separate the processing code and the interface so that they can be checked properly.
> 
> What do you mean? Even if we are using Shiny, it is impossible to include figures generated from interactive mode, we have to put figures generated from Shiny outside vignettes. Am I wrong? Or if I use Shiny, figures generated from Shiny can be included in vignettes automatically? I am very confused. What is improved if we employ Shiny instead of menu?

Sorry I wasn't clear, I will try to be more specific. I think:

- Using Shiny would give a nicer interface to the user
- Using Shiny would allow you to better separate the interactive and functional code. It's probably still possible to do that for other approaches though.
- No, using Shiny won't allow you to include plots directly in the vignette but separating the plotting and visualisation code would. You could then show an example of a plot without going through the whole selection process.
- Showing the plots I think is a secondary issue, it would be fine to include screenshots of the app (or whatever interface), but maybe you don't need as many as you have now.
- In the end how to do this is up to you. We are just trying to provide suggestions that we think will lead to a better experience for the user and a more maintainable package for you.

Also, I am the main reviewer for your package. I am sure `@LiNk-NY` is happy to provide input if needed but please stop pinging them in every message.

**Comment by tagtag** [comment]
@lazappi 
> > > I had not realised the batch mode is not functional and only there to pass the tests. This is something I think we want to avoid. It could be very confusing to the user and means the functions are not tested properly. I am beginning to think that Shiny is the best option in this case. That would give a nicer interface to the user than the menus and would make it easier to separate the processing code and the interface so that they can be checked properly.
> > 
> > 
> > What do you mean? Even if we are using Shiny, it is impossible to include figures generated from interactive mode, we have to put figures generated from Shiny outside vignettes. Am I wrong? Or if I use Shiny, figures generated from Shiny can be included in vignettes automatically? I am very confused. What is improved if we employ Shiny instead of menu?
> 
> Sorry I wasn't clear, I will try to be more specific. I think:
> 
> * Using Shiny would give a nicer interface to the user
> * Using Shiny would allow you to better separate the interactive and functional code. It's probably still possible to do that for other approaches though.
> * No, using Shiny won't allow you to include plots directly in the vignette but separating the plotting and visualisation code would. You could then show an example of a plot without going through the whole selection process.
> * Showing the plots I think is a secondary issue, it would be fine to include screenshots of the app (or whatever interface), but maybe you don't need as many as you have now.
> * In the end how to do this is up to you. We are just trying to provide suggestions that we think will lead to a better experience for the user and a more maintainable package for you.
> 
> Also, I am the main reviewer for your package. I am sure `@LiNk-NY` is happy to provide input if needed but please stop pinging them in every message.


Without including interactive mode, I cannot demonstrate downstream analysis, since only after the selections through interactive mode,  we can have the input for downstream analysis. The users needs to investigate plot one by one, to perform selection.  It is impossible to generate batch mode to provide results without interactive mode. Do you understand the problem? You said that batch mode without doing anything is not allowed. But intearctive mode is not allowed in vignettes. But we need results form inteerctive mode. If dummy bacth mode is not allowed, I cannot generate vignettes at all.
Did you agree with including dummy batch mode now? Or  are you still opposed to this? In this case, please let me know the acceptable way by which we can get the result from interactive mode without either performing interactive mode or  including dummy batch mode that simply provides the result when interactive mode was performed. Did you understand the points? I need the answer on this point.

**Comment by tagtag** [comment]
@lazappi This error seemed to come from the system, not from mine. I am now asking this issue in Slack. 
https://community-bioc.slack.com/archives/C34NC134G/p1679156164253489

**Comment by tagtag** [comment]
@lazappi I have stopped employing dummy batch mode and follow your advice. I have found some solution on how to avoid   fig.keep="none". I will inform you when I address all the concerns you raised as much as possible.

**Comment by tagtag** [comment]
Dear @lazappi ,

Thanks for your detailed comments. I tried to address them. Although my package shows Error, It is not because of my side but system side (see https://community-bioc.slack.com/archives/C34NC134G/p1679197255196319?thread_ts=1679156164.253489&cid=C34NC134G). 
As you can see below, it has passed in Mac.
![image](https://user-images.githubusercontent.com/11694497/226368310-d7c15984-001d-4e8b-8b11-885a4c00d02e.png)
Thus, it might be OK for you to comment revised one. If not, I will wait for the system probelm is solved.


> 
> Please address the comments as best as you can. When you are ready for me to check the package again please reply to let me know with a summary of changes you have made or any other responses. You can use the "Quote reply" option in the `...` menu on this comment to reply directly to my points below.
> 
> Luke
> 
> # Review
> **Key:** 🚨 Required ⚠️ Recommended 🟢 Optional ❓ Question
> 
> ## General comments
> * [ ]  ❓ The functions in the package seem to mostly be small helper functions that build on the functionality in **{TDbasedUFE}**. Have you considered adding them to the existing package rather than creating a new one?

Yes, I have considered adding them to the existing package. However, it will be too huge to install. I might add more functions to it. In addition to this, the functions in this package and those in future packaged might not be required by many peoplr. Thus, I think that it mght b better to have this one as a separated one.

> 
> ## DESCRIPTION file
> * [ ]  ⚠️ It may be appropriate to add "Software" to the `biocViews` field

I have added.

> * [ ]  🟢 It is recommended to add a `BugReports` field, this is usually a link to the GitHub issues page or the Bioconductor support site

I have added. 

> * [ ]  🟢 It is recommended to add a `URL` field, this is usually a link to the GitHub repository
> 

I have added. 

> ## NAMESPACE file
> * [ ]  🚨 `PrepareSummarizedExperimentTensorRect` should start with a lower case letter to match the other functions

I did.

> * [ ]  🚨 Please use selective imports using `importFrom` instead of importing all with `import`

I did.

> ## The NEWS file
> * [ ]  ⚠️ Please keep the NEWS file up to date with the latest changes

I did.

> ## Package data
> * [ ]  🚨 The data in `inst/extdata/` should be documented in `inst/script`

I did. 
> ## Documentation
> * [ ]  🟢 There were quite a lot of typos in different places, the **{spelling}** package can help with finding these

I did.
> 
> ### README
> * [ ]  🚨 The `README` looks incomplete, please update to fill in the missing content

Although I added some, I am not sure what to add.

> * [ ]  ⚠️ It doesn't look as though you are using any RMarkdown functionality, if so it will be simpler to use a plain `README.md` file instead of `README.Rmd`

I did.

> 
> ### Vignette
> * [ ]  🚨 Please expand the introduction in the main vignette to describe the package in more detail. This should include the motivation for including the package in Bioconductor and comparisions to any similar packages (if applicable).

I did.

> * [ ]  ⚠️ It was not clear to me which order to look at the vignettes. I expected to start with `TDbasedUFEadv.Rmd` but that only convered technical detail, not any information about how to use the package. I suggest renaming/reorganising the vignettes to make this clearer.

It is renamed as "Explanation_of_TDbasedUFEadv"

> * [ ]  ⚠️ It also wasn't clear to me what the difference was between `Quickstart.Rmd` and `Quickstart2.Rmd`. I would suggest that this can be combined.

They were merged and renamed as "How_to_use_TDbasedUFEadv"

> * [ ]  🚨 Whatever you decide to do I suggest using a more specific file name than "QuickStart.Rmd", sometimes it causes issues if different packages have vignettes with the same name

I did as above.

> * [ ]  🚨 The main vignette should have an installation section that includes the Bioconductor installation instructions. These chunk should **not** be evaluated.

I did.

> * [ ]  ⚠️ There are lots of images in the vignettes directory, are all of these required or can some be produced by the vignette (I think this is related to the point about interactivity below)?

Figs were removed.

> * [ ]  ⚠️ I don't think you need to show all the menu items, I think it would be enough to mention there is an interactive option and maybe show one or two screenshots

Figs were removed.

> * [ ]  ⚠️ Please don't use `fig.keep="none"` if it can be avoided, in general the vignette should show the same output as the user would see if they ran the code themselves

Now most figures generated inside vignettes. Some conceptual figures remain there.

> * [ ]  🟢 Usually it is better to use `library(package)` rather than `require(package)`. I also find it easier to follow the code if the library calls are in a separate chunk at the start.

I did.

> 
> ### Man pages
> * [ ]  🚨 You have included the word "Title" in the function documentation title which is then rendered in the final help page, either use `@title TEXT` or just `TEXT` on the first line

I did.

> * [ ]  🟢 Some of the variables in the examples are not very descriptive (e.g. `dummy`) which makes them a bit hard to understand

I renamed them.

> 
> ## Unit tests
> * [ ]  🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks

I did.

> 
> ## Code
> ### R
> * [ ]  🚨 Please use named indicies instead of numeric indices where possible

I did as much as possible, but I found only a few.

> * [ ]  ⚠️ I found the `SummarizedExperimentTensorRect` class a bit unclear in a few ways
>   
>   * [ ]  Given the name I expected it to inherit from `SummarizedExperiment` but that is not the case, I would suggest using a more descriptive name

As denoted in the above, I am now planning to  add more related packages in the future (in a few years?).  I would like to let them have similar names. Thus, it would be better for them to have general names not specific ones.

>   * [ ]  It doesn't seem to be used by the user, how necessary is it to create a class rather than just using a list?

In future, I plan to GRange, which is included in class but not used now.  In this case, class will be better.

>   * [ ]  It is defined in a file called `init.R` which is not clear, usually this file is called something like `AllClasses.R` or the name of the class

I did. 

> * [ ]  ⚠️ Please consider using more descriptive argument names, some of them (e.g. "k", "j", "LIST", "L") do not clearly describe what they are used for

I did.

> * [ ]  ⚠️ Consider adding checks for whether valid inputs are provided to function arguments

I did, although they might not be proper ones that you intended.

> * [ ]  ⚠️ I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`

I am not sure. In my ability, I could not. Do you have any specific suggestion for `for' loops?

> * [ ]  ⚠️ Generally functions should not output plots as side effects, generally it is better to have separate analysis and plotting functions

Interactive mode is required in this implementation.  They must be  executed in batch mode in vignettes and must generate figures in vignettes. Please let them to generate figures.


> * [ ]  ⚠️ It may be better to separate interactive functions for the non-interactive versions (or change how the options work), I found it a bit surprising that interactive was the default mode. Maybe one function could produce a plot that the user uses to select a parameter which they then pass to another function?

Functions in interactive mode must generate figures.  Otherwise, no interactive function can work.


> * [ ]  🟢 This is mostly personal preference but I found your code quite hard to read because of the lack of indenting and whitespace. You may want to look at the [tidyverse style guide](https://style.tidyverse.org/) or the `styler` package.

I have reformed them using  `styler` package as suggested.

> 
> ## Shiny apps
> * [ ]  ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?

I have removed menu and used Shiny app. However, it generated the usage of "<<-" which is noted. Shiny app has not ability to return value, thus I could not avoid the usage of "<<-" to deliver the values gotten in interactive mode to downstream analyses.  

I am not very sure if you are satisfied with this revision. If not, I will try to improve it further.

Thanks for your efforts again.

**Comment by tagtag** [comment]
Dear @lazappi ,

Although build system problem was not yet solved, since the present version  [0.99.14](http://bioconductor.org/spb_reports/TDbasedUFEadv_buildreport_20230321065303.html)  passed check in Mac, I am glad if you can review the present version to which I have replied to [your comments](https://github.com/Bioconductor/Contributions/issues/2944#issuecomment-1470764456)  at [issuecomment-1476367945](https://github.com/Bioconductor/Contributions/issues/2944#issuecomment-1476367945).

Yours, tag.

**Comment by tagtag** [comment]
Dear @lazappi ,

Now Single Package Builder problem was fixed and the present version, [0.99.15](http://bioconductor.org/spb_reports/TDbasedUFEadv_buildreport_20230321182334.html), has gotten "OK" flag. I am glad if you can comment the present version when you have time.

Yours, tag.

**Comment by lazappi** [comment]
> > * [ ]   🚨 Please use selective imports using `importFrom` instead of importing all with `import`
> 
> I did.

There are still quite a few normal imports. It is ok if these are unavoidable but I don't think this is the case here. I think some of them (**{BiocStyle}**, **{RTCGA}**,...) could be made suggested dependencies as they are only used for the vignette/examples, not the core package functionality. 

> > ## The NEWS file
> > 
> > * [ ]   ⚠️ Please keep the NEWS file up to date with the latest changes
> 
> I did.

I can't see any changes to the `NEWS.md` file. I just see:

```
# TDbasedUFEadv 0.99.9

* Addressing many comments from reviewers. 
```

> > ## Package data
> > 
> > * [ ]   🚨 The data in `inst/extdata/` should be documented in `inst/script`
> 
> I did.

It's great you added a link to where to get the data but this should also include some information about what the data is, file formats, information it contains, what it is used for etc.

> > ### README
> > 
> > * [ ]   🚨 The `README` looks incomplete, please update to fill in the missing content
> 
> Although I added some, I am not sure what to add.

I think this is ok now

> > * [ ]   ⚠️ It was not clear to me which order to look at the vignettes. I expected to start with `TDbasedUFEadv.Rmd` but that only convered technical detail, not any information about how to use the package. I suggest renaming/reorganising the vignettes to make this clearer.
> 
> It is renamed as "Explanation_of_TDbasedUFEadv"

I think this is better now, it is clear that one vignette explains the method and the other one explains how to use the package. Just double check you have updated the names everywhere, I think I still saw some references to "Quickstart"

> > * [ ]   🟢 Usually it is better to use `library(package)` rather than `require(package)`. I also find it easier to follow the code if the library calls are in a separate chunk at the start.
> 
> I did.

The examples still use `require()`.

> > ## Unit tests
> > 
> > * [ ]   🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks
>
> I did.

There are still no actual tests here. As you are using the **{testthat}** package your tests should be wrapped in the `test_that()` function. Please refer to the [**{testthat}** documentation](https://testthat.r-lib.org/) or the testing chapter of the [R packages book](https://r-pkgs.org/testing-basics.html).

> > * [ ]   ⚠️ I found the `SummarizedExperimentTensorRect` class a bit unclear in a few ways
> >   
> >   * [ ]   Given the name I expected it to inherit from `SummarizedExperiment` but that is not the case, I would suggest using a more descriptive name
> 
> As denoted in the above, I am now planning to add more related packages in the future (in a few years?). I would like to let them have similar names. Thus, it would be better for them to have general names not specific ones.
>
> > * [ ]   It doesn't seem to be used by the user, how necessary is it to create a class rather than just using a list?
> 
> In future, I plan to GRange, which is included in class but not used now. In this case, class will be better.

I don't think this makes sense and it is confusing for users to use one of the standard Bioconductor object names without inheriting from that object. It is hard to anticipate what kind of object you might need in the future so I would suggest focusing on what you need now. If you are extending the `SummarizedExperiment` object then this name is ok, if not please use a more specific name to avoid confusion.

> > * [ ]   ⚠️ Consider adding checks for whether valid inputs are provided to function arguments
> 
> I did, although they might not be proper ones that you intended.

I think these are a good start. This can be difficult so it is usually expanded/improved over time.
 
> > * [ ]   ⚠️ I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`
> 
> I am not sure. In my ability, I could not. Do you have any specific suggestion for `for' loops?

Some of these could maybe be written using `lapply()`, for example:

```r
  for (i in seq_len(length(Cancer_cell_lines_p)))
  {
    expDrug <- rbind(expDrug, data.frame(Cancer_cell_lines_p[[i]]))
  }
 ```

Could be written something like (untested so maybe need some adjusting):

```r
  for (i in seq_len(length(Cancer_cell_lines_p)))
  {
    expDrug <- rbind(expDrug, data.frame(Cancer_cell_lines_p[[i]]))
  }
  
 expDrug <- lapply(seq_along(Cancer_cell_lines_p), function(i) {
     data.frame(Cancer_cell_lines_p[[i]])
)
expDrug <- do.call("rbind", expDrug)
```

`for` loops aren't inherently bad though as long as you pre-allocate memory before which I think you do in most cases. Other cases such as when you are making multiple plots are unavoidable so that's fine. Overall, I don't think you need to worry about this too much.

As a note, you can use `seq_along(x)` rather than `seq_len(length(x))`. You can also use `10L` rather than `as.integer(10)`

> > ## Shiny apps
> > 
> > * [ ]   ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?
> 
> I have removed menu and used Shiny app. However, it generated the usage of "<<-" which is noted. Shiny app has not ability to return value, thus I could not avoid the usage of "<<-" to deliver the values gotten in interactive mode to downstream analyses.
> 
> I am not very sure if you are satisfied with this revision. If not, I will try to improve it further.

I think this works much better now! I like that you can just specify a value or use the app to select one based on the plots. I have a few minor suggestions. It would maybe be good to add some text to the app describing what to look for in the plots. It might also be easier if the user could see all the plots at once, at the moment I don't think you can go back if you think an earlier one looked better. Can the user look at the plots separately from the selection app? I think that would also be useful, particularly for things like including them in a report.

**Comment by tagtag** [comment]
Dear @lazappi ,
 
> > > ## Unit tests
> > > 
> > > * [ ]    🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks
> > 
> > 
> > I did.
> 
> There are still no actual tests here. As you are using the **{testthat}** package your tests should be wrapped in the `test_that()` function. Please refer to the [**{testthat}** documentation](https://testthat.r-lib.org/) or the testing chapter of the [R packages book](https://r-pkgs.org/testing-basics.html).

I am not sure what you mean. I  wrote in testthat.R as

library(testthat)
library(TDbasedUFEadv)

test_check("TDbasedUFEadv")

and in testthat directory I placed

test-TDbasedUFadv.R

Thus test-TDbasedUFadv.R should be executed. This structure is similar to others, for example,

https://code.bioconductor.org/browse/AlpsNMR/tree/master/tests/

I am glad if you can advise me about what the problem is.

Yours, tag.

**Comment by tagtag** [comment]
Dear @lazappi ,

Thanks for your quick and detailed comments. In the below I tried to address your concerns.

> > > * [ ]    🚨 Please use selective imports using `importFrom` instead of importing all with `import`
> > 
> > 
> > I did.
> 
> There are still quite a few normal imports. It is ok if these are unavoidable but I don't think this is the case here. I think some of them (**{BiocStyle}**, **{RTCGA}**,...) could be made suggested dependencies as they are only used for the vignette/examples, not the core package functionality.
> 

I did. RTCGA cannot be moved to suggested, since a function convertTCGA was used in some functions.

> > > ## The NEWS file
> > > 
> > > * [ ]    ⚠️ Please keep the NEWS file up to date with the latest changes
> > 
> > 
> > I did.
> 
> I can't see any changes to the `NEWS.md` file. I just see:
> 
> ```
> # TDbasedUFEadv 0.99.9
> 
> * Addressing many comments from reviewers. 
> ```


I modified it. I hope that it is OK now.

> > > ## Package data
> > > 
> > > * [ ]    🚨 The data in `inst/extdata/` should be documented in `inst/script`
> > 
> > 
> > I did.
> 
> It's great you added a link to where to get the data but this should also include some information about what the data is, file formats, information it contains, what it is used for etc.
> 

I have added more information.

> 
> > > * [ ]    ⚠️ It was not clear to me which order to look at the vignettes. I expected to start with `TDbasedUFEadv.Rmd` but that only convered technical detail, not any information about how to use the package. I suggest renaming/reorganising the vignettes to make this clearer.
> > 
> > 
> > It is renamed as "Explanation_of_TDbasedUFEadv"
> 
> I think this is better now, it is clear that one vignette explains the method and the other one explains how to use the package. Just double check you have updated the names everywhere, I think I still saw some references to "Quickstart"

I am not sure which one you mean, but a mention about "Quickstart" is not for this package, but the previous one, TDbasedUFE. 

> 
> > > * [ ]    🟢 Usually it is better to use `library(package)` rather than `require(package)`. I also find it easier to follow the code if the library calls are in a separate chunk at the start.
> > 
> > 
> > I did.
> 
> The examples still use `require()`.


I am so sorry. Now all require()s are cheged to library().
 
> 
> > > ## Unit tests
> > > 
> > > * [ ]    🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks
> > 
> > 
> > I did.
> 
> There are still no actual tests here. As you are using the **{testthat}** package your tests should be wrapped in the `test_that()` function. Please refer to the [**{testthat}** documentation](https://testthat.r-lib.org/) or the testing chapter of the [R packages book](https://r-pkgs.org/testing-basics.html).

I have commented [here](https://github.com/Bioconductor/Contributions/issues/2944#issuecomment-1479367733).
> 
> > > * [ ]    ⚠️ I found the `SummarizedExperimentTensorRect` class a bit unclear in a few ways
> > >   
> > >   * [ ]    Given the name I expected it to inherit from `SummarizedExperiment` but that is not the case, I would suggest using a more descriptive name
> > 
> > 
> > As denoted in the above, I am now planning to add more related packages in the future (in a few years?). I would like to let them have similar names. Thus, it would be better for them to have general names not specific ones.
> > > * [ ]    It doesn't seem to be used by the user, how necessary is it to create a class rather than just using a list?
> > 
> > 
> > In future, I plan to GRange, which is included in class but not used now. In this case, class will be better.
> 
> I don't think this makes sense and it is confusing for users to use one of the standard Bioconductor object names without inheriting from that object. It is hard to anticipate what kind of object you might need in the future so I would suggest focusing on what you need now. If you are extending the `SummarizedExperiment` object then this name is ok, if not please use a more specific name to avoid confusion.

I renamed it to `TensorRect`.

> 
> > > * [ ]    ⚠️ Consider adding checks for whether valid inputs are provided to function arguments
> > 
> > 
> > I did, although they might not be proper ones that you intended.
> 
> I think these are a good start. This can be difficult so it is usually expanded/improved over time.
> 

I see.

> > > * [ ]    ⚠️ I think some of the `for` loops could maybe be replaced with other iteration methods, e.g. `vapply` or `lapply`
> > 
> > 
> > I am not sure. In my ability, I could not. Do you have any specific suggestion for `for' loops?
> 
> Some of these could maybe be written using `lapply()`, for example:
> 
> ```r
>   for (i in seq_len(length(Cancer_cell_lines_p)))
>   {
>     expDrug <- rbind(expDrug, data.frame(Cancer_cell_lines_p[[i]]))
>   }
> ```
> 
> Could be written something like (untested so maybe need some adjusting):
> 
> ```r
>   for (i in seq_len(length(Cancer_cell_lines_p)))
>   {
>     expDrug <- rbind(expDrug, data.frame(Cancer_cell_lines_p[[i]]))
>   }
>   
>  expDrug <- lapply(seq_along(Cancer_cell_lines_p), function(i) {
>      data.frame(Cancer_cell_lines_p[[i]])
> )
> expDrug <- do.call("rbind", expDrug)
> ```
> 
> `for` loops aren't inherently bad though as long as you pre-allocate memory before which I think you do in most cases. Other cases such as when you are making multiple plots are unavoidable so that's fine. Overall, I don't think you need to worry about this too much.


I have changed this one. I could not find any others that can be removed.

> As a note, you can use `seq_along(x)` rather than `seq_len(length(x))`. You can also use `10L` rather than `as.integer(10)`

I did.

> 
> > > ## Shiny apps
> > > 
> > > * [ ]    ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?
> > 
> > 
> > I have removed menu and used Shiny app. However, it generated the usage of "<<-" which is noted. Shiny app has not ability to return value, thus I could not avoid the usage of "<<-" to deliver the values gotten in interactive mode to downstream analyses.
> > I am not very sure if you are satisfied with this revision. If not, I will try to improve it further.
> 
> I think this works much better now! I like that you can just specify a value or use the app to select one based on the plots. I have a few minor suggestions. It would maybe be good to add some text to the app describing what to look for in the plots. It might also be easier if the user could see all the plots at once, at the moment I don't think you can go back if you think an earlier one looked better. Can the user look at the plots separately from the selection app? I think that would also be useful, particularly for things like including them in a report.

Although I added one sentence, it does not make sense much, since what should be selected is context dependent. This is the reason why I need an interactive mode. 
Showing all figures at once is not a good idea, since the number of figure vary. And this process, showing figures one by one, was proposed by some experimentalist who will use this package. He hates to see all figures at once, since it is simply annoying. Actually, it is very likely that selection takes place within the first a few figures. 
Figures can be shown, since now batch mode is not dummy but actually generate figures because of your suggestions. Thus users can do as
pdf(file="fig.pdf")
(execute batch mode)
dev.off()
and they can include figures into report. 

I am not sure if you are satisfied with the present version yet.

**Comment by lazappi** [comment]
> > > > ## The NEWS file
> > > > 
> > > > * [ ]     ⚠️ Please keep the NEWS file up to date with the latest changes
> > > 
> > > 
> > > I did.
> > 
> > 
> > I can't see any changes to the `NEWS.md` file. I just see:
> > ```
> > # TDbasedUFEadv 0.99.9
> > 
> > * Addressing many comments from reviewers. 
> > ```
> 
> I modified it. I hope that it is OK now.

Yes, this is good, thanks! Please keep this up to date as you make changes in the future.

> > > > * [ ]     ⚠️ It was not clear to me which order to look at the vignettes. I expected to start with `TDbasedUFEadv.Rmd` but that only convered technical detail, not any information about how to use the package. I suggest renaming/reorganising the vignettes to make this clearer.
> > > 
> > > 
> > > It is renamed as "Explanation_of_TDbasedUFEadv"
> > 
> > 
> > I think this is better now, it is clear that one vignette explains the method and the other one explains how to use the package. Just double check you have updated the names everywhere, I think I still saw some references to "Quickstart"
> 
> I am not sure which one you mean, but a mention about "Quickstart" is not for this package, but the previous one, TDbasedUFE.

In `Enrichement.Rmd` Line 38. If this refers to another package you might want to add a HTML link to the vignette.
 
> > > > ## Unit tests
> > > > 
> > > > * [ ]     🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks
> > > 
> > > 
> > > I did.
> > 
> > 
> > There are still no actual tests here. As you are using the **{testthat}** package your tests should be wrapped in the `test_that()` function. Please refer to the [**{testthat}** documentation](https://testthat.r-lib.org/) or the testing chapter of the [R packages book](https://r-pkgs.org/testing-basics.html).
> 
> I have commented [here](https://github.com/Bioconductor/Contributions/issues/2944#issuecomment-1479367733).

Yes, I saw that comment and replied to it. As I said before you need to use the `test_that()` function to turn the code you have into actual tests. What you have now will be run but it is not a test recognised by the build/checking systems. Please check the documentation I linked to above.

This is the one major point left. If you fix this I think I should be able to accept the package.

> > As a note, you can use `seq_along(x)` rather than `seq_len(length(x))`. You can also use `10L` rather than `as.integer(10)`
> 
> I did.

There are still a few `as.integer()` calls that can be replaced in the same way.

> > > > ## Shiny apps
> > > > 
> > > > * [ ]     ❓ I thought the plan was to replace the menus with a Shiny app, is this still the case?
> > > 
> > > 
> > > I have removed menu and used Shiny app. However, it generated the usage of "<<-" which is noted. Shiny app has not ability to return value, thus I could not avoid the usage of "<<-" to deliver the values gotten in interactive mode to downstream analyses.
> > > I am not very sure if you are satisfied with this revision. If not, I will try to improve it further.
> > 
> > 
> > I think this works much better now! I like that you can just specify a value or use the app to select one based on the plots. I have a few minor suggestions. It would maybe be good to add some text to the app describing what to look for in the plots. It might also be easier if the user could see all the plots at once, at the moment I don't think you can go back if you think an earlier one looked better. Can the user look at the plots separately from the selection app? I think that would also be useful, particularly for things like including them in a report.
> 
> Although I added one sentence, it does not make sense much, since what should be selected is context dependent. This is the reason why I need an interactive mode. Showing all figures at once is not a good idea, since the number of figure vary. And this process, showing figures one by one, was proposed by some experimentalist who will use this package. He hates to see all figures at once, since it is simply annoying. Actually, it is very likely that selection takes place within the first a few figures. Figures can be shown, since now batch mode is not dummy but actually generate figures because of your suggestions. Thus users can do as pdf(file="fig.pdf") (execute batch mode) dev.off() and they can include figures into report.
> 
> I am not sure if you are satisfied with the present version yet.

I think what you have now is ok, so I am happy with how it is. I do think it could be improved in a few ways though but you can look at that after the package is accepted.

**Comment by tagtag** [comment]
Dear @lazappi 

Thanks for your comments, but I still cannot understand the problem in test.


> Yes, I saw that comment and replied to it. As I said before you need to use the `test_that()` function to turn the code you have into actual tests. What you have now will be run but it is not a test recognised by the build/checking systems. Please check the documentation I linked to above.
> 
> This is the one major point left. If you fix this I think I should be able to accept the package.

I compare mine with [AlpsNMR](https://code.bioconductor.org/browse/AlpsNMR/tree/master/), for example. 

In AlpsNMR, tests directory is as 
![image](https://user-images.githubusercontent.com/11694497/227517434-346e41a2-7236-499e-9a3f-8aaeb51e066e.png)
and TDbasedUFadv is  
![image](https://user-images.githubusercontent.com/11694497/227517653-d9d756ef-fca3-4de6-a60f-aef2e12ed86c.png)
Thus they are the same. 

For testthat.R AlpsNMR

library(testthat)
library(AlpsNMR)

test_check("AlpsNMR")

for TDbasedUFEadv

library(testthat)
library(TDbasedUFEadv)

test_check("TDbasedUFEadv")

This the same again.

In testthat directory,  for AlpsNMR
![image](https://user-images.githubusercontent.com/11694497/227518207-eb5c7d0c-13bf-4cb7-9e39-94f352760142.png)
for TDbasedUFEadv
![image](https://user-images.githubusercontent.com/11694497/227518349-28538221-778f-4cb2-99d5-be57fdbb70be.png)

The only difference is that I have only one file and AlpsNMR has multiple file. What you think the problem is that I have only one file that does not corresponds to individual files in R directory?

If so, there is a reason why I could not prepare the files each of which corresponds to individual files in R directory.
Most of functions in R directory cannot be tested independently, In order to executing A.R requires execute B.R before executing A.R.  Thus no test file that test only A.R. Thus we placed one file including all functions in R.
 
Please advice me how to solve this difficulty if you mean that I have to have multiple files each of which corresponds to individual files in R directory. If you mean something different please let me know. Although I read the document you listed carefully, I cannot understand the problem at all.

Please advise me what to do more practically.

**Comment by tagtag** [comment]
@lazappi I understand the problem for test. You mean that I have to use test_that{} function in test-TDbasedUFEadv, I think.

**Comment by tagtag** [comment]
Dear @lazappi ,

Thank you very much for you comments. In the bellow, I tied to address your comments,


> In `Enrichement.Rmd` Line 38. If this refers to another package you might want to add a HTML link to the vignette.

Done. 


> 
> > > > > ## Unit tests
> > > > > 
> > > > > * [ ]      🚨 Please add unit tests for all exported functions. There is some code in the tests file currently by no actual expectation checks
> > > > 
> > > > 
> > > > I did.
> > > 
> > > 
> > > There are still no actual tests here. As you are using the **{testthat}** package your tests should be wrapped in the `test_that()` function. Please refer to the [**{testthat}** documentation](https://testthat.r-lib.org/) or the testing chapter of the [R packages book](https://r-pkgs.org/testing-basics.html).
> > 
> > 
> > I have commented [here](https://github.com/Bioconductor/Contributions/issues/2944#issuecomment-1479367733).
> 
> Yes, I saw that comment and replied to it. As I said before you need to use the `test_that()` function to turn the code you have into actual tests. What you have now will be run but it is not a test recognised by the build/checking systems. Please check the documentation I linked to above.
> 
> This is the one major point left. If you fix this I think I should be able to accept the package.
> 
I have used test_this() in test-TDbasedUFEadv. 
There is only one file, but it was accepted in the [previous submission](https://code.bioconductor.org/browse/TDbasedUFE/tree/master/tests/testthat/).

> 
> There are still a few `as.integer()` calls that can be replaced in the same way.
> 
Done. 

 I am not very sure if you are satisfied with this revision. If not, I will try to improve it further.

**Comment by lazappi** [comment]
> Dear @lazappi
> 
> Thanks for your comments, but I still cannot understand the problem in test.
> 
> > Yes, I saw that comment and replied to it. As I said before you need to use the `test_that()` function to turn the code you have into actual tests. What you have now will be run but it is not a test recognised by the build/checking systems. Please check the documentation I linked to above.
> > This is the one major point left. If you fix this I think I should be able to accept the package.
> 
> I compare mine with [AlpsNMR](https://code.bioconductor.org/browse/AlpsNMR/tree/master/), for example.
> 
> In AlpsNMR, tests directory is as ![image](https://user-images.githubusercontent.com/11694497/227517434-346e41a2-7236-499e-9a3f-8aaeb51e066e.png) and TDbasedUFadv is ![image](https://user-images.githubusercontent.com/11694497/227517653-d9d756ef-fca3-4de6-a60f-aef2e12ed86c.png) Thus they are the same.
> 
> For testthat.R AlpsNMR
> 
> library(testthat) library(AlpsNMR)
> 
> test_check("AlpsNMR")
> 
> for TDbasedUFEadv
> 
> library(testthat) library(TDbasedUFEadv)
> 
> test_check("TDbasedUFEadv")
> 
> This the same again.
> 
> In testthat directory, for AlpsNMR ![image](https://user-images.githubusercontent.com/11694497/227518207-eb5c7d0c-13bf-4cb7-9e39-94f352760142.png) for TDbasedUFEadv ![image](https://user-images.githubusercontent.com/11694497/227518349-28538221-778f-4cb2-99d5-be57fdbb70be.png)
> 
> The only difference is that I have only one file and AlpsNMR has multiple file. What you think the problem is that I have only one file that does not corresponds to individual files in R directory?
> 
> If so, there is a reason why I could not prepare the files each of which corresponds to individual files in R directory. Most of functions in R directory cannot be tested independently, In order to executing A.R requires execute B.R before executing A.R. Thus no test file that test only A.R. Thus we placed one file including all functions in R.
> 
> Please advice me how to solve this difficulty if you mean that I have to have multiple files each of which corresponds to individual files in R directory. If you mean something different please let me know. Although I read the document you listed carefully, I cannot understand the problem at all.
> 
> Please advise me what to do more practically.

I can see that you are using the `test_that()` function now. I'm not sure if you have added that since I last checked or if I missed that earlier (apologies if I did). This is ok but it should be split up into several smaller tests. At the moment if any of these fail you will not be able to tell which check is the issue and where to look in your code. It is not encouraged but it is possible to have code outside the `test_that()` function. So you would have something more like this:

```r
data(dataset1)

test_that("function 1 works", {
	result <- function1(dataset1)
	expect_indentical(result, expected_result)
})

test_that("function 2 works", {
	output1 <- function1(dataset1)
	result <- function2(output1)
	expect_indentical(result, expected_result)
})
```

The problem with adding code outside of `test_that()` is that if there is an error the tests will stop completely whereas if it is inside the function the testing package will handle that. It is ok to have multiple `expect_` functions inside a test and this is encouraged to test different parts of the output.

The other important thing is that the name of the test (the first text input to the function) is clear so that it is obvious what the test is checking. This is shown in the output if something fails so it is very helpful for knowing which part of the code to look at.

**Comment by tagtag** [comment]
Dear @lazappi ,

Thank you very much for you prompt replies and comments.

> I can see that you are using the `test_that()` function now. I'm not sure if you have added that since I last checked or if I missed that earlier (apologies if I did). This is ok but it should be split up into several smaller tests. At the moment if any of these fail you will not be able to tell which check is the issue and where to look in your code. It is not encouraged but it is possible to have code outside the `test_that()` function. So you would have something more like this:
> 
> ```r
> data(dataset1)
> 
> test_that("function 1 works", {
> 	result <- function1(dataset1)
> 	expect_indentical(result, expected_result)
> })
> 
> test_that("function 2 works", {
> 	output1 <- function1(dataset1)
> 	result <- function2(output1)
> 	expect_indentical(result, expected_result)
> })
> ```
> 
> The problem with adding code outside of `test_that()` is that if there is an error the tests will stop completely whereas if it is inside the function the testing package will handle that. It is ok to have multiple `expect_` functions inside a test and this is encouraged to test different parts of the output.
> 
> The other important thing is that the name of the test (the first text input to the function) is clear so that it is obvious what the test is checking. This is shown in the output if something fails so it is very helpful for knowing which part of the code to look at.

I have tried to fragment test_that() so as to correspond to individual functions, but I was also forced to include many "<<-"s  which are advised to avoid as much as possible. However, these functions are tightly interrelated, it was impossible to avoid the usage of "<<-"s. 
I am glad if you can accept this version. 

Yours, tag.

**Comment by lazappi** [comment]
> > I can see that you are using the `test_that()` function now. I'm not sure if you have added that since I last checked or if I missed that earlier (apologies if I did). This is ok but it should be split up into several smaller tests. At the moment if any of these fail you will not be able to tell which check is the issue and where to look in your code. It is not encouraged but it is possible to have code outside the `test_that()` function. So you would have something more like this:
> > ```r
> > data(dataset1)
> > 
> > test_that("function 1 works", {
> > 	result <- function1(dataset1)
> > 	expect_indentical(result, expected_result)
> > })
> > 
> > test_that("function 2 works", {
> > 	output1 <- function1(dataset1)
> > 	result <- function2(output1)
> > 	expect_indentical(result, expected_result)
> > })
> > ```  
> >   
> > The problem with adding code outside of `test_that()` is that if there is an error the tests will stop completely whereas if it is inside the function the testing package will handle that. It is ok to have multiple `expect_` functions inside a test and this is encouraged to test different parts of the output.
> > The other important thing is that the name of the test (the first text input to the function) is clear so that it is obvious what the test is checking. This is shown in the output if something fails so it is very helpful for knowing which part of the code to look at.
> 
> I have tried to fragment test_that() so as to correspond to individual functions, but I was also forced to include many "<<-"s which are advised to avoid as much as possible. However, these functions are tightly interrelated, it was impossible to avoid the usage of "<<-"s. I am glad if you can accept this version.

Thanks, this is much better! Each test now checks a single thing and it's easy to see what that is. I think you should be able to avoid using `<<-`. Some of these aren't necessary because the assigned either isn't used again or is overwritten before it is used. Others could be replaced by either randomly generated data or by pre-computing values and saving them as part of the package. This would also help make the tests independent of each other.

My preference would be for this to be addressed but I don't think it is strictly necessary for the package to be accepted so please let me know what you plan to do.

**Comment by tagtag** [comment]
Dear @lazappi,

Thank you very much for you quick reply.

> > I have tried to fragment test_that() so as to correspond to individual functions, but I was also forced to include many "<<-"s which are advised to avoid as much as possible. However, these functions are tightly interrelated, it was impossible to avoid the usage of "<<-"s. I am glad if you can accept this version.
> 
> Thanks, this is much better! Each test now checks a single thing and it's easy to see what that is. I think you should be able to avoid using `<<-`. Some of these aren't necessary because the assigned either isn't used again or is overwritten before it is used. Others could be replaced by either randomly generated data or by pre-computing values and saving them as part of the package. This would also help make the tests independent of each other.
> 
> My preference would be for this to be addressed but I don't think it is strictly necessary for the package to be accepted so please let me know what you plan to do.

 I have checked one by one the usage of "<<-"s in test  because of your advises  and erased some of them. I don't think that the remaining  (see below) can be either removed  or replaced with random numbers since they are class objects that include more than simple numbers.

L10 expDrug -> used later in L14
L20 Z -> used later in  L34
L49 SVD -> L61 ->  used later in  L66
L56 Z ->  used later in  L65
L80 Z ->  used later in L84
L84, L96 Z -> used later in L108

Saving them as pre-computing values is not a good idea either, since they are often huge and might be beyond the individual files size limit of Bioconductor package (5MB). 
Thus,  we cannot removed  more "<<-"s, I think. 

I am glad if you can accept the present version. 

Yours, tag.

**Comment by lazappi** [comment]
> Dear @lazappi,
> 
> Thank you very much for you quick reply.
> 
> > > I have tried to fragment test_that() so as to correspond to individual functions, but I was also forced to include many "<<-"s which are advised to avoid as much as possible. However, these functions are tightly interrelated, it was impossible to avoid the usage of "<<-"s. I am glad if you can accept this version.
> > 
> > 
> > Thanks, this is much better! Each test now checks a single thing and it's easy to see what that is. I think you should be able to avoid using `<<-`. Some of these aren't necessary because the assigned either isn't used again or is overwritten before it is used. Others could be replaced by either randomly generated data or by pre-computing values and saving them as part of the package. This would also help make the tests independent of each other.
> > My preference would be for this to be addressed but I don't think it is strictly necessary for the package to be accepted so please let me know what you plan to do.
> 
> I have checked one by one the usage of "<<-"s in test because of your advises and erased some of them. I don't think that the remaining (see below) can be either removed or replaced with random numbers since they are class objects that include more than simple numbers.
> 
> L10 expDrug -> used later in L14 L20 Z -> used later in L34 L49 SVD -> L61 -> used later in L66 L56 Z -> used later in L65 L80 Z -> used later in L84 L84, L96 Z -> used later in L108
> 
> Saving them as pre-computing values is not a good idea either, since they are often huge and might be beyond the individual files size limit of Bioconductor package (5MB). Thus, we cannot removed more "<<-"s, I think.
> 
> I am glad if you can accept the present version.
> 
> Yours, tag.

I will approve the package because there is some testing and I think you have made a big effort to address all my comments. I think you are likely to run into issues with this approach though so I encourage you to look into alternatives.

**Comment by lazappi** [comment]
Congratulations, **{TDbasedUFEadv}** has been accepted into Bioconductor 🎉! It can take a few days to be picked up by the build system but then it should be available in Bioconductor devel and the next Bioconductor release.

**Comment by tagtag** [comment]
Dear @lazappi ,

Thank you very much for your great efforts to improve my package. I will try to improve this further as much as possible.

Yours, tag.

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/tagtag.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("TDbasedUFEadv")`. The package 'landing page' will be created at

https://bioconductor.org/packages/TDbasedUFEadv

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by tagtag** [comment]
@lshep Thank you very much for your very quick job in spite of your very busy schedule during release period!

**Comment by lshep** [comment]
Thank you for your contributions to Bioconductor!
